src/python-common/ceph/deployment/service_spec.py

```diff
@@ -140,17 +140,111 @@ def validate(self) -> None:
         assert_valid_host(self.hostname)
 
 
+HostPatternType = Union[str, None, Dict[str, Union[str, bool, None]], "HostPattern"]
+
+
+class HostPattern():
+    def __init__(self, pattern: Optional[str] = None, is_regex: bool = False) -> None:
+        self.pattern = pattern
+        self.is_regex = is_regex
```

Comment:
A thought: it might be good o try and run `re.compile` on a regex pattern up front and cache it for a possible minor performance improvement. The former makes sure the pattern is a valid regex and the later could save a small amount of cpu if  the are a lot of host patterns to match in a large cluster.

Reply:
I don't think performance will be of much importance here, but I've added an additional `compiled_regex` attribute that is populated and used if the constructor gets a pattern of type regex